external/fv3fit/fv3fit/pytorch/cyclegan/modules.py

```diff
@@ -0,0 +1,235 @@
+import logging
+
+from typing import Callable, Literal, Protocol, Union
+import torch.nn as nn
+import torch
+
+logger = logging.getLogger(__name__)
+
+
+def relu_activation(**kwargs):
+    def relu_factory():
+        return nn.ReLU(**kwargs)
+
+    return relu_factory
+
+
+def tanh_activation():
+    return nn.Tanh()
+
+
+def leakyrelu_activation(**kwargs):
+    def leakyrelu_factory():
+        return nn.LeakyReLU(**kwargs)
+
+    return leakyrelu_factory
+
+
+def no_activation():
+    return nn.Identity()
+
+
+class ConvolutionFactory(Protocol):
+    def __call__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: int,
+        padding: Union[str, int] = 0,
+        output_padding: int = 0,
+        stride: int = 1,
+        stride_type: Literal["regular", "transpose"] = "regular",
+        bias: bool = True,
+    ) -> nn.Module:
+        """
+        Create a convolutional layer.
+
+        Layer takes in and returns tensors of shape [batch, tile, channels, x, y].
+
+        Args:
+            in_channels: number of input channels
+            out_channels: number of output channels
+            kernel_size: size of the convolution kernel
+            padding: padding to apply to the input, should be an integer or "same"
+            output_padding: argument used for transpose convolution
+            stride: stride of the convolution
+            stride_type: type of stride, one of "regular" or "transpose"
+            bias: whether to include a bias vector in the produced layers
+        """
+        ...
+
+
+class CurriedModuleFactory(Protocol):
+    def __call__(self, in_channels: int, out_channels: int) -> nn.Module:
+        """
+        Create a torch module.
+
+        Args:
+            in_channels: number of input channels
+            out_channels: number of output channels
+        """
+        ...
+
+
+class FoldTileDimension(nn.Module):
+    """
+    Module wrapping a module which takes [batch, channel, x, y] data into one
+    which takes [batch, tile, channel, x, y] data by folding the tile dimension
+    into the batch dimension.
+    """
+
+    def __init__(self, wrapped):
+        super(FoldTileDimension, self).__init__()
+        self._wrapped = wrapped
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        x = inputs.reshape(-1, *inputs.shape[2:])
+        x = self._wrapped(x)
+        return x.reshape(inputs.shape[0], inputs.shape[1], *x.shape[1:])
+
+
+def single_tile_convolution(
```

Comment:
Abstracting convolution with a factory like this will allow me to swap single tile convolution for graph-based analogous implementations through configuration only, while leaving the model building code unchanged.